code/lib/core-webpack/src/to-importFn.ts

```diff
@@ -4,6 +4,20 @@ import { globToRegexp } from '@storybook/core-common';
 
 import { importPipeline } from './importPipeline';
 
+function adjustRegexToExcludeNodeModules(originalRegex: RegExp) {
+  const originalRegexString = originalRegex.source;
+  const startsWithCaret = originalRegexString.startsWith('^');
+  const excludeNodeModulesPattern = startsWithCaret ? '(?!.*node_modules)' : '^(?!.*node_modules)';
+
+  // Combine the new exclusion pattern with the original regex
+  const adjustedRegexString = startsWithCaret
+    ? `^${excludeNodeModulesPattern}${originalRegexString.substring(1)}`
+    : excludeNodeModulesPattern + originalRegexString;
+
+  // Create and return the new regex
+  return new RegExp(adjustedRegexString);
+}
```

Comment:
Would it make sense to instead of doing this adjustment here to do it in `normalizeStoriesEntry`? 

https://github.com/storybookjs/storybook/blob/c98bddd2f2244a3a869c8e680e0e6734a03ebec3/code/lib/core-common/src/utils/normalize-stories.ts#L35

I guess I'm just wondering where else we use the stories entry to lookup or watch files? For instance, I think we do it here also:

https://github.com/storybookjs/storybook/blob/c98bddd2f2244a3a869c8e680e0e6734a03ebec3/code/lib/core-server/src/utils/watch-story-specifiers.ts#L26-L31

Reply:
So given your comment below, we might not want to entirely disregard stories from `node_modules` for indexing / watching the index. Or at least if we did we'd want to commit to stories from the `node_modules` not being supported at all in SB.

I think we should definitely consider that (for perf reasons). I can imagine that setting up file watchers recursively in `node_modules` as `watch-story-specificers` may very well be doing could be *very* costly. At the least we could feature flag it so it's off by default.

I would say we do that work separate to this PR, but the thing is if we were to adjust the stories entry, we wouldn't need to make the other changes that are in this PR.

Reply:
> Would it make sense to instead of doing this adjustment here to do it in `normalizeStoriesEntry`?
> 
> https://github.com/storybookjs/storybook/blob/c98bddd2f2244a3a869c8e680e0e6734a03ebec3/code/lib/core-common/src/utils/normalize-stories.ts#L35
> 
> I guess I'm just wondering where else we use the stories entry to lookup or watch files? For instance, I think we do it here also:
> 
> https://github.com/storybookjs/storybook/blob/c98bddd2f2244a3a869c8e680e0e6734a03ebec3/code/lib/core-server/src/utils/watch-story-specifiers.ts#L26-L31

@tmeasday If you don't mind, I would like to pair on this since I am not sure how easily achievable this is. `normalizeStoriesEntry` doesn't deal with regex, but with globs and having excluding patterns in a glob string is not that easy. We could change `normalizeStoriesEntry` completely to not return a glob string for `files`, but instead a e.g. picomatch `matcher`, which can be configured to ignore certain patterns: https://github.com/micromatch/picomatch. But this would require major refactorings.